common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/CertBasedAuthFactory.java

```diff
@@ -45,18 +45,21 @@ public class CertBasedAuthFactory {
     public CertBasedAuthFactory(@NonNull final Activity activity) {
```

Comment:
Wondering if we can have 3 managers.
1. USB.
2. NFC.
3. On-device.

Reply:
with this, we wouldn't need startUsbDiscovery() and startNfcDiscovery() - just startDiscovery() would suffice.

Reply:
Splitting up the SmartcardCertBasedAuthManager into two makes sense to me, and correspondingly, I'll split up the SmartcardCertBasedAuthChallengeHandler into two as well. The USB manager will remain in CertBasedAuthFactory, but I think the NFC manager can be created in UserChoiceCertBasedAuthChallengeHandler, since that's where the first NFC discovery will always start. USB discovery needs to remain active for the duration of the WebView being shown, while NFC discovery will only be on whenever we explicitly ask the user to tap.

In response to the second part of the feedback, the design you're proposing is similar to the initial sample I showed to the UX designer, and it was rejected in favor of having that additional dialog where the user first chooses between general smartcard or on-device. Choosing the smartcard option will bring the user to a dialog like 1. described above, but with a cancel button. The UI flow as shown in the video I attached is what the designer has approved.

Also, as I recall in my previous refactoring PR, we were trying to keep control of the dialogs (via DialogHolder) separate from CertBasedAuthFactory. I also think that, since there are at least 3 dialogs involved with the process of choosing a type of CBA, it's been helpful to have a separate challenge handler to take care of those dialogs and then route the ClientCertRequest to a more specific challenge handler.

Reply:
I don't see any attached video. Would you mind sharing the link?

Reply:
I'll send thru chat

Reply:
actually never mind. just saw it in the PR review chat room.

Reply:
After watching the video... here's my feedback.

- We probably want to handle the flow selection in this class. Once it's decided, there is no way for the user to change it. This means that the first dialog, the "please tap" dialog and the cancel dialog (or any UI that would let the user switch between the flow) should probably be handled here.  

- createCertBasedAuthChallengeHandler might be turned into a callback - which eventually trigger one of the three flows, USB, NFC, On-Device. Maybe it could perform asynchronous polling to check if any devices (of any manufacturers, not just yubikey) is detected, then proceed with that manager.

With the current structure, I do see some problems,  i.e.
- What would happen if the user plugs in a non-NFC yubikey when we ask the user to tap?
- What if we need to support more than YubiKey?

Reply:
To answer the two concerns:
- So I think I've tested this scenario in the past, but testing it now, when plugging in the key, the plugging in acts as a tap lol. I believe what I had before (but somehow got lost) was an if statement that only runs the connection callback if  `isNfc` parameter is true. The result of that would be if the user plugs in a non-NFC yubikey while the user is already on the NFC path, the consent dialog pops up, but nothing else should happens, and the user can unplug and still proceed with authentication via NFC. 
- The manager classes will hold lists of the smartcard-specific managers and will control which discoveries are on and off. I believe the USB discovery method of YubiKit will only call its callback after identifying the connected device as a YubiKey. I'm not as certain for NFC, so that's something I'll need to follow up with Yubico with. 
Best case scenario, supposing other companies would do the same thing for both USB and NFC, we would be able to keep all discoveries on in the "please tap" dialog, make a note of the initial manager that made this connection, and then use this manager when it comes to reading future information off the smartcard. 
If the other smartcard brands cannot or refuse to implement a vendor_id check, and supposing we also won't be able to implement that ourselves for NFC... then we would either need some sort of hint from eSTS (not certain this would be possible) or (more likely) I will need to create an additional dialog that asks the user which smartcard vendor they're working with. 

In terms of moving the code in UserChoiceCertBasedAuthChallengeHandler to CertBasedAuthFactory, I understand logically why we would do this, so despite my hesitancy with introducing UI as well as telemetry (what I'm working on rn) into it, I'll make the changes.